storage: improve snapshot info messages in log truncation

tbg opened the pull request:
It was warning excessively, in particular during restore operations,
where we constantly truncate the log to the quorum commit index due
to the Raft log being over target size. We have "approximate" in-order
delivery (or Raft doesn't work very well) for MsgApp, and so we assume
that inflight messages will reach the trailing followers which would
thus not need a snapshot.

Shelled out from #32594.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/32672)
<!-- Reviewable:end -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32672#-:-LSQ5fvL0wHRCopuOtWh:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

tbg committed:
storage: improve snapshot info messages in log truncation

It was warning excessively, in particular during restore operations,
where we constantly truncate the log to the quorum commit index due
to the Raft log being over target size. We have "approximate" in-order
delivery (or Raft doesn't work very well) for MsgApp, and so we assume
that inflight messages will reach the trailing followers which would
thus not need a snapshot.

Release note: None

tbg commented:
bors r=petermattis
